src/libPMacc/include/functor/Filtered.hpp

```diff
@@ -0,0 +1,255 @@
+/* Copyright 2017 Rene Widera
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc_types.hpp"
+#include "filter/Interface.hpp"
+#include "mappings/threads/WorkerCfg.hpp"
+
+#include <string>
+
+
+namespace PMacc
+{
+namespace functor
+{
+namespace acc
+{
+
+
+    /** interface to combine a filter and a functor on the accelerator
+     *
+     *
+     * @tparam T_FilterOperator PMacc::filter::operators, type concatenate the
+     *                          results of the filter
+     * @tparam T_Filter PMacc::filter::Interface, type of the filter
+     * @tparam T_Functor PMacc::functor::Interface, type of the functor
+     */
+    template<
+        typename T_FilterOperator,
+        typename T_Filter,
+        typename T_Functor
+    >
+    struct Filtered :
+        private T_Filter,
+        private T_Functor
+    {
+        using Filter = T_Filter;
+        using Functor = T_Functor;
+
+        DINLINE Filtered(
+            Filter const & filter,
+            Functor const & functor
+        ) :
+            Filter( filter ),
+            Functor( functor )
+        {
+
+        }
+
+        /** execute the functor depending of the filter result
+         *
+         * Call the filter for each argument and if the combined result is true
+         * the user functor is called.
+         *
+         * @param args arguments passed to the functor
+         *             - the filter is executed for each argument
+         *             - the functor get passed all arguments
+         */
+        template< typename ... T_Args >
+        DINLINE auto operator( )(  T_Args && ... args )
+        -> void
+        {
+            if(
+                T_FilterOperator{ }(
+                    ( *reinterpret_cast< Filter * >( this ) )( args ) ...
+                )
+            )
+                ( *reinterpret_cast< Functor * >( this ) )( args ... );
+        }
+    };
```

Comment:
Keep what you have now, but for future reference to myself, this is what this operator() implements:

```C++
// needed since filter is not implemented as a member var to save 1 byte
auto filter = *reinterpret_cast< Filter * >( this );
T_FilterOperator filterOperator{};

// call the filter on each argument and combine the results
bool const combinedResult = filterOperator( filter( args ) ... );

if( combinedResult )
{
    // needed since functor is not implemented as a member var to save 1 byte
    auto functor = *reinterpret_cast< Functor * >( this );
    functor( args ... );
}
```